lib/ReactViews/Generic/Timer/drawTimer.js

```diff
@@ -0,0 +1,186 @@
+"use strict";
+
+import { arc as d3Arc } from "d3-shape";
+import { select as d3Select } from "d3-selection";
+import { interpolate as d3Interpolate } from "d3-interpolate";
+import { easeLinear as d3EaseLinear } from "d3-ease";
+
+import defined from "terriajs-cesium/Source/Core/defined";
+
+/**
+ * Returns a generator function that returns a string representing the svg path of an arc.
+ * @param {number} radius
+ */
+function arcGenerator(radius) {
+  return d3Arc()
+    .innerRadius(0)
+    .outerRadius(radius)
+    .startAngle(0);
+}
+
+// Interpolate from 0 to 2*pi radians
+function angleInterpolator(t, startAngle = 0) {
+  return d3Interpolate(startAngle, Math.PI * 2)(t);
+}
+
+/**
+ * Runs the timer animation, making an arc fill from 0 to 100% of the circle.
+ * @param {number} radius Radius of timer.
+ * @param {number} interval Timer duration in seconds.
+ * @param {DOMElement} elapsedTimeElement SVG path containing the elapsed time "pie".
+ * @param {DOMElement} backgroundElement SVG path containing the background circle.
+ * @param {number} [options.deltaOpacity=0.5] Change in opacity when fading in and out timer.
+ * @param {number} [options.opacityAnimationInterval=3] How long fade in/out lasts in seconds.
+ * @param {number} [options.minOpacity=0.1] When fading out, doesn't let opacity go below `minOpacity`.
+ * @param {number} [options.elapsed=0] How much time has already passed.
+ */
+function animateTimer(
+  radius,
+  interval,
+  elapsedTimeElement,
+  backgroundElement,
+  options = {}
+) {
+  options = {
+    deltaOpacity: defined(options.deltaOpacity) ? options.deltaOpacity : 0.7,
+    opacityAnimationInterval: defined(options.opacityAnimationInterval)
+      ? options.opacityAnimationInterval
+      : 3,
+    minOpacity: defined(options.minOpacity) ? options.minOpacity : 0.1,
+    elapsed: defined(options.elapsed) ? options.elapsed : 0
+  };
+
+  // The arc representing the elapsed time should be filled up to the current time.
+  // We find the elapsed time as a percentage of the total duration, and then get the angle interpolator to calculate
+  // the corresponding angle.
+  const startAngle = angleInterpolator(options.elapsed / interval);
+  elapsedTimeElement
+    .datum({ endAngle: angleInterpolator(startAngle) })
+    .transition("arc" + new Date().getTime().toString())
+    .duration((interval - options.elapsed) * 1000) // d3 uses milliseconds
+    .ease(d3EaseLinear)
+    // attrTween requires a generator function A that returns a function B
```

Comment:
a bit confusing to call the function passed into `attrTween` a "generator function" in the comment as the generator function has some specific meaning. In this case it is just a function that returns another function? Maybe currying is a more accurate description?